opencog/atoms/core/ImplicationLink.cc

```diff
@@ -61,26 +61,23 @@ ImplicationLink::ImplicationLink(Link &l)
 void ImplicationLink::extract_variables(const HandleSeq& oset)
 {
 	size_t sz = _outgoing.size();
-	// For now we assume no implicit scoping, so the number of
-	// arguments is 2 when sugar syntax isn't used (no scoping), or 3
-	// when the sugar syntax (scoping) is used.
+
 	if (sz < 2 or 3 < sz)
 		throw InvalidParamException(TRACE_INFO,
-			"Expecting an outgoing set size of at most two, got %d", sz);
+			"Expecting an outgoing set of size 2 or 3, got %d", sz);
+
+	// If there is an explicit variable declaration, then use that.
+	// If there is no variable declaration, then assume that all of
+	// the free variables in the link are scoped.
+	ScopeLink::extract_variables(oset);
```

Comment:
OK -- now, to explain MapLink and how it currently interacts with ImplicationLink.  

MapLink can be thought of in three different ways:
* It is like "map" in scheme, haskell -- it applies a "function"  to a list or set.  (that's why its called MapLink)
* It is like UnPutLink, so that if "function f" is a pattern with variables in it, it can extract values for those variables.  Thus, its kind-of-like GetLink, except that GetLink searches the entire atomspace, but MapLink only searches the set/list that is it's argument.  You can see why that would be kind-of nice.
* MapLink combined with ImplicationLink is kind-of-like BindLink, but instead of working on the entire atomspace, it only works on the argument list/set. 

I hope the above makes it clear why MapLink is fun, useful, interesting. 

The MapLink unit test fails on this third, last thingy -- combining Map+Implication.   There is very little code that uses/needs this, but I believe that the AIML converter (or maybe the OpenPsi subset of it?) does depend on this working correctly.  So if we break that, the chatbot would break. 

However, there are alternatives. Note that BindLink is "exactly" equal to a GetLink followed by a PutLink.  We could "get rid of" BindLink, its kind of a historical accident that happens to be used a lot.  Recall that BindLink was originally created as a kind-of ImplicationLink that could be evaluated "on the spot".

Similarly, the combination of MapLink+ImplicationLink could probably be replaced by PutLink+MapLink.   That is, just like Put+Get is a kind-of-implication acting on all of atomspace, similarly, Put+Map is a kind-of-implication acting only on its input set/list.  Both of these   perform a re-write of P(x)->Q(x)  where BindLink searches the entire atomspace for x, while Put+Map searches only its input set/list for x.

This we could do this: nuke ImplicationLink.cc entirely, update the MapLinkUTet to use Put+Map instead, and update OpenPsi/AIML to use Put+Map instead of Map+Implication.

We could also create a new link type, e.g. ReWriteLink, that performs the P(x)->Q(x) operation.  We could also re-define MapLink so that if it has multiple args, the first arg would be P(x) and the second arg would be Q(x).   Any of these approaches seem plausible to me.

At any rate, I do not see why we need ImplicationLink.cc at all.  It seems to cause semantic problems.